### PR TITLE
RFC: realtime v2

### PR DESCRIPTION
**Pull request by mansueli**
## What kind of change does this PR introduce?

Bug fix, feature, docs update, ...

## What is the current behavior?

Please link any relevant issues here.

## What is the new behavior?

Feel free to include screenshots if it includes visual changes.

## Additional context

Add any other context or screenshots.

### CODE REVIEW

**Comment by silentworks** [comment]
@mansueli this PR will need to be updated as some other PRs got merged in which are now causing merge conflicts with this one.

File: poetry.lock [hunk 1]
```diff
@@ -1,15 +1,14 @@
-# This file is automatically @generated by Poetry 1.8.2 and should not be changed by hand.
-
+# This file is automatically @generated by Poetry 1.7.1 and should not be changed by hand.
```
Comment (juancarlospaco) [inline poetry.lock:1]: This file is autogenerated, but it is replacing a newer gen tool with an older one, 
latest Poetry is `1.8.3` https://github.com/python-poetry/poetry/releases/tag/1.8.3
maybe is best to always move forwards in semver not backwards IMHO.

File: realtime/exceptions.py [hunk 2]
```diff
@@ -8,3 +8,15 @@ def __init__(self, func_name: str):
...
         return f"A WS connection has not been established. Ensure you call Socket.connect() before calling Socket.{self.offending_func_name}()"
+
+
+class AuthorizationError(Exception):
+    """
+    Raised when there is an authorization failure for private channels
+    """
+
+    def __init__(self, message: str = "Authorization failed for private channel"):
+        self.message: str = message
```
Comment (juancarlospaco) [inline realtime/exceptions.py:19]:
```suggestion
    def __init__(self, message: str = None):
        self.message: str = message or "Authorization failed for private channel"
```

Mutable default argument. IMHO.

**Review by juancarlospaco — APPROVED** [review]
👍

File: example/app.py [hunk 3]
```diff
@@ -19,17 +18,27 @@ async def realtime(payload):
     print("async realtime ", payload)
 
 if __name__ == "__main__":
-    ID = os.getenv("SUPABASE_ID")
+    ID = 'unjmcdnawyleplexbxja'#os.getenv("SUPABASE_ID")
```
Comment (juancarlospaco) [inline example/app.py:21]: Do not commit personal credentials and IDs.

File: example/app.py [hunk 4]
```diff
@@ -19,17 +18,27 @@ async def realtime(payload):
...
     URL = f"https://{ID}.supabase.co"
-    JWT = os.getenv("API_KEY")
+    JWT = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InVuam1jZG5hd3lsZXBsZXhieGphIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTcxOTkzNDY2NCwiZXhwIjoyMDM1NTEwNjY0fQ.qL9CBx9wUEdUWTRjxJTEIhUhnHDeeexu3UHiLqxqQ7o'#os.getenv("API_KEY")
```
Comment (juancarlospaco) [inline example/app.py:23]: Ditto.

File: example/app.py (hunk 3 above)
Comment (mansueli) [reply example/app.py:21]: This is a test project, but point taken.

**Review by J0 — APPROVED** [review]

**Comment by Hawpe** [comment]
Is this a working example? I can't get things going using the example file. I've tried a combination of things but I don't seem to ever receive a payload. I'm testing this locally, subscribing to a channel in Python:

```python
import time
from realtime.channel import Channel
from realtime.connection import Socket


def broadcast_callback(payload):
    print("broadcast: ", payload)


URL = "ws://127.0.0.1:54321"
JWT = "..."


broad_socket = Socket(URL, JWT, auto_reconnect=True)
broadcast_channel = Channel(broad_socket, topic="broadcast-test")
broadcast_channel.on_broadcast("test", callback=broadcast_callback).subscribe()
time.sleep(120)
```

And sending a broadcast message from a NodeJS app:
```ts
supabase.channel('broadcast-test').send({
    type: 'broadcast',
    event: 'test',
    payload: { message: 'test message' }
});
```

I've tried setting the socket and channel parameters to use `{"broadcast": {"self": True}}` and sending the broadcast from the Python code too, didn't work either. I've confirmed that sending and receiving works fine on the NodeJS side.

I've tried listening to the packets and I sometimes see this message:
`{"ref":null,"event":"phx_reply","payload":{"status":"error","response":{"reason":"unmatched topic"}},"topic":"broadcast-test"}`
In other tests I've been able to get rid of the message, the payload was properly sent and I still didn't get anything in the callback.

Am I missing something?

**Pull request merged**